File: src/net/eduard/api/lib/game/Calculator.java
package net.eduard.api.lib.game;

import java.util.Map;

import net.eduard.api.lib.modules.Extra;
import net.eduard.api.lib.storage.Storable;
import net.eduard.api.lib.storage.Storable.*;

@StorageAttributes(inline = true)
public class Calculator implements Storable {
	private double base = 10;
	private double perLevel = 5;
	private double limit = 0;
	private double variationPerLevel = 0;
	private double variation = 0;

	public Calculator() {

	}

	public Calculator(double base, double perLevel) {
		super();
		this.perLevel = perLevel;
		this.base = base;
	}

	public Calculator(double base, double perLevel, double limit) {
		super();
		this.limit = limit;
		this.perLevel = perLevel;
		this.base = base;
	}

	public double getValue(double level) {
		double result = (perLevel * (level)) + base;
		double variationNumber = variation + (variationPerLevel * level);
		double random = Extra.getRandomDouble(-variationNumber, variationNumber);
		double resultFinal = result + random; // + extra;
		if (limit > 0 && resultFinal > limit) {
			return limit;
		}
		if (resultFinal <= 0) {
			resultFinal = 0;
		}
		return resultFinal;
	}

	public double getBase() {
		return base;
	}

	public void setBase(double base) {
		this.base = base;
	}

	public double getPerLevel() {
		return perLevel;
	}

	public void setPerLevel(double perLevel) {
		this.perLevel = perLevel;
	}

//	public double getVariationBonus() {
//		return variationBonus;
//	}
//
//	public void setVariationBonus(double variationBonus) {
//		this.variationBonus = variationBonus;
//	}

	public double getLimit() {
		return limit;
	}

	public void setLimit(double limit) {
		this.limit = limit;
	}

	public double getVariationPerLevel() {
		return variationPerLevel;
	}

	public void setVariationPerLevel(double variationPerLevel) {
		this.variationPerLevel = variationPerLevel;
	}

	public double getVariation() {
		return variation;
	}

	public void setVariation(double variation) {
		this.variation = variation;
	}

//	public double getVariationOnLevels() {
//		return variationOnLevels;
//	}
//
//	public void setVariationOnLevels(double variationOnLevels) {
//		this.variationOnLevels = variationOnLevels;
//	}
}

File: src/main/java/net/eduard/api/lib/game/Calculator.kt
package net.eduard.api.lib.game

import net.eduard.api.lib.storage.annotations.StorageAttributes


@StorageAttributes(inline = true)
class Calculator(
    var base: Double,
    var perLevel: Double,
    var limit: Double,
    var variationBetweenLevels: Double,
    var variationAditional: Double
) {
    companion object {
        fun linearByBasePercent(base: Double, basePercent: Double): Calculator {
            return Calculator(base, base * basePercent)
        }
    }

    constructor() : this(10.0, 5.0, 0.0)

    constructor(
        base: Number = 10,
        perLevel: Number = 5,
        limit: Number = 0,
        variationBetweenLevels: Number = 0,
        variationAditional: Number = 0
    ) : this(
        base.toDouble(),
        perLevel.toDouble(),
        limit.toDouble(),
        variationBetweenLevels.toDouble(),
        variationAditional.toDouble()
    )

    constructor(
        base: Number = 10,
        perLevel: Number = 5,
        limit: Number = 0,
    ) : this(base.toDouble(), perLevel.toDouble(), limit.toDouble(), 0, 0)

    constructor(
        base: Number = 10,
        perLevel: Number = 5
    ) : this(base.toDouble(), perLevel.toDouble(), 0, 0, 0)

    fun getValue(level: Double): Double {
        val result = perLevel * level + base
        var variationPower = 0
        if (variationBetweenLevels > 0 && level > 0)
            variationPower = (level / variationBetweenLevels).toInt()
        val variationTotal = variationPower * variationAditional
        var resultFinal = result + variationTotal
        if (limit > 0 && resultFinal > limit) {
            return limit
        }
        if (resultFinal <= 0) {
            resultFinal = 0.0
        }
        return resultFinal
    }

}